qiskit/transpiler/passes/routing/swap_strategies/commuting_2q_gate_router.py

```diff
@@ -0,0 +1,347 @@
+    of logical qubits 0, 1, 2, and 3.
+
+    .. code-block:: python
+
+        from qiskit import QuantumCircuit
+        from qiskit.opflow import PauliSumOp
+        from qiskit.circuit.library import PauliEvolutionGate
+        from qiskit.transpiler import Layout, CouplingMap, PassManager
+        from qiskit.transpiler.passes import FullAncillaAllocation
+        from qiskit.transpiler.passes import EnlargeWithAncilla
+        from qiskit.transpiler.passes import ApplyLayout
+        from qiskit.transpiler.passes import SetLayout
+
+        from qiskit.transpiler.passes.routing.swap_strategies import (
+            SwapStrategy,
+            FindCommutingPauliEvolutions,
+            Commuting2qGateRouter,
+        )
+
+        # Define the circuit on logical qubits
+        op = PauliSumOp.from_list([("IZZI", 1), ("ZIIZ", 2), ("ZIZI", 3)])
+        circ = QuantumCircuit(4)
+        circ.append(PauliEvolutionGate(op, 1), range(4))
+
+        # Define the swap strategy on qubits before the initial_layout is applied.
+        swap_cmap = CouplingMap(couplinglist=[(0, 1), (1, 2), (2, 3)])
+        swap_strat = SwapStrategy(swap_cmap, swap_layers=[[(0, 1), (2, 3)], [(1, 2)]])
+
+        # Chose qubits 0, 1, 3, and 4 from the backend coupling map shown above.
+        backend_cmap = CouplingMap(couplinglist=[(0, 1), (1, 2), (1, 3), (3, 4)])
+        initial_layout = Layout.from_intlist([0, 1, 3, 4], *circ.qregs)
+
+        pm_pre = PassManager(
+            [
+                FindCommutingPauliEvolutions(),
+                Commuting2qGateRouter(swap_strat),
+                SetLayout(initial_layout),
+                FullAncillaAllocation(backend_cmap),
+                EnlargeWithAncilla(),
+                ApplyLayout(),
+            ]
+        )
+
+        # Insert swap gates, map to initial_layout and finally enlarge with ancilla.
+        pm_pre.run(circ).draw("mpl")
+
+    This pass manager relies on the ``current_layout`` which corresponds to the qubit layout as
+    swap gates are applied. The pass will traverse all nodes in the dag. If a node should be
+    routed using a swap strategy then it will be decomposed into sub-instructions with swap
+    layers in between and the ``current_layout`` will be modified. Nodes that should not be
+    routed using swap strategies will be added back to the dag taking the ``current_layout``
+    into account.
+    """
+
+    def __init__(self, swap_strategy: Optional[SwapStrategy] = None) -> None:
+        """
+        Args:
+            swap_strategy: An instance of a :class:`.SwapStrategy` that holds the swap layers
+                that are used, and the order in which to apply them, to map the instruction to
+                the hardware. If this field is not given if should be contained in the
+                property set of the pass. This allows other passes to determine the most
+                appropriate swap strategy at run-time.
+        """
```

Comment:
what happens if I initialize this pass with some "wrong" swap strategy? i.e. can i make this produce a wrong circuit by supplying a swap strategy that does not properly route the circuit?

Reply:
If you initialize with a swap strategy that does not reach the required connectivity you get an error. E.g. if I take the example https://github.com/Qiskit/qiskit-terra/blob/fdce1881dfb55259aa4238b2d5645997541dbd51/qiskit/transpiler/passes/routing/commuting_2q_gate_routing/commuting_2q_gate_router.py#L55-L94 and change the swap strategy to
```python
swap_cmap = CouplingMap(couplinglist=[(0, 1), (1, 2), (2, 3)])
swap_strat = SwapStrategy(swap_cmap, (((0, 1), (2, 3)), ))
```
which does not reach the required connectivity I get the error
```
TranspilerError: 'SwapStrategy with swap layers:\n((0, 1), (2, 3)),\non [[0, 1], [1, 2], [2, 3]] coupling map. cannot implement all edges in {(1, 2), (0, 3), (1, 3)}.'
```

Also if you try and create a swap strategy that does not fit the coupling map you get an error, e.g.:
```python
swap_cmap = CouplingMap(couplinglist=[(0, 1), (1, 2), (2, 3)])
swap_strat = SwapStrategy(swap_cmap, (((0, 1), (2, 3)), ((0, 2), )))  # (0, 2) is not valid
```